fabio/test/codecs/test_edfimage1.py

```diff
@@ -0,0 +1,263 @@
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+# Unit tests
+Reading edf files as originally specified by expg with fabio.module edfimage.py
+
+07/06/2019 PB
+"""
+from __future__ import print_function, with_statement, division, absolute_import
+import unittest
+import os
+import numpy
+import shutil
+import io
+import logging
+
+logger = logging.getLogger(__name__)
+
+import fabio
+from fabio.edfimage import edfimage
+from ..utilstest import UtilsTest
+
+#logging.basicConfig(level=logging.DEBUG)
+#logging.basicConfig(level=logging.WARNING)
+#logging.basicConfig(level=logging.INFO)
+#logging.debug('This will get logged')
+#logging.warning('This will get logged')
+#logging.info('This will get logged')
+
+#==========================================
+def fopen(filename=None,frameno=None):
+    header=None
+    data=None
+    frame=None
+    logging.debug("fopen(filename={},frameno={})".format(filename,frameno))
+    if filename:
+        image=fabio.open(filename)
+        if frameno is None:
+            frameno=0
+
+        nframes=image.nframes
+
+        #if image.classname == "EdfImage":
+        if hasattr(image,"npsdframes"):
+            npsdframes=image.npsdframes
+            nerrorframes=image.nerrorframes
+        else:
+            npsdframes=nframes
+            nerrorframes=0
+
+        if frameno >= npsdframes:
+            logging.warning("Psd frame {} out of range: 0 <= {} < {}".format(frameno,frameno,npsdframes))
+
+        if frameno <0:
+            if -frameno > nerrorframes:
+                logging.warning("Error frame {} out of range: {} <= {} < 0 ".format(frameno,-nerrorframes,frameno))
+            frameno += nframes
+
+        if frameno in range(nframes):
+            if nframes>1:
+                frame = image.getframe(frameno)
+                data = frame.data
+                header=frame.header
+            else:
+                # Single frame
+                data = image.data
+                header= image.header
+            frame=fabio.fabioimage.FabioFrame(data=data, header=header)
+        else:
+            raise IOError("fopen: Cannot access frame: {} (0<=frame<{})".format(frameno, nframes))
+
+    return(frame)
+
+#============================================
+
+ROOT = None
+
+def test_00(self,filename,avglist=None,keylist=None):
+    """
+    Checks the shape of the data arrays.
+    Compares the mean value of each data frame array, with the number given in avglist.
+    filename: name of file to test
+    avglist: list of average values for each frame, stops, if not equal
+    keylist: list of keys to read, stops, if not available
+    If avglist or keylist is shorter than frameno, the last value in the list is used.
+    """
+    global ROOT
+    filename = os.path.join(ROOT, filename)
```

Comment:
Using global variables is BAD. You should rewrite your function using a method and share this ROOT.

Reply:
done